### PR TITLE
Fix build without DRM/GBM and simple build/crash fixes

### PR DESCRIPTION
**Pull request by blino**
Hi,

This fixes a few build and crash issues, notably for embedded environments (no DRM/GBM, stripped xkbcommon installation).

### CODE REVIEW

**Comment by blino** [comment]
Actually, it could be better to introduce a GBM flag, both for code and cmake, to avoid duplicating if WPE_BACKEND(DRM) || WPE_BACKEND(WAYLAND) checks everywhere.

**Comment by blino** [comment]
Maybe have separate WPE_VIEW_BACKEND and WPE_RENDERING_BACKEND ?

**Comment by zdobersek** [comment]
Right, it should be possible to use the Wayland protocol with different cross-process buffer management APIs. GBM is just one of them.

So yeah, this needs work.

I'd keep the WPE_BACKEND_DRM and WPE_BACKEND_WAYLAND flags. To address this I would introduce e.g. WPE_BUFFER_MANAGEMENT_GBM, and guard GBM-specific code with that. Other API usages could be guarded with WPE_BUFFER_MANAGEMENT_WHATEVER, (or rather WPE_BUFFER_MANAGEMENT(WHATEVER) in the actual code).

**Comment by blino** [comment]
I have updated this to have a WPE_BUFFER_MANAGEMENT_GBM macro.
This allows to build without gbm, and prepare for other buffer management types.
Though, the gbm code in ViewBackendWayland.cpp should probably be factorized out at some point.

I have tested that this still builds fine with gbm, which is still enabled by default.

Thanks for your feedback.

File: Source/WebCore/PlatformWPE.cmake [hunk 1]
```diff
@@ -307,6 +302,25 @@ list(APPEND WebCore_INCLUDE_DIRECTORIES
     ${WEBP_INCLUDE_DIRS}
 )
 
+if (USE_WPE_BUFFER_MANAGEMENT_GBM)
```
Comment (zdobersek) [inline Source/WebCore/PlatformWPE.cmake]: This block shouldn't be necessary. Source/WebCore/platform/graphics/gbm doesn't exist anymore, and we shouldn't include gbm.h anymore from WebCore.

File: Source/WebCore/PlatformWPE.cmake [hunk 2]
```diff
@@ -307,6 +302,25 @@ list(APPEND WebCore_INCLUDE_DIRECTORIES
...
+    list(APPEND WebCore_INCLUDE_DIRECTORIES
+        "${WEBCORE_DIR}/platform/graphics/gbm"
+        ${LIBGBM_INCLUDE_DIRS}
+    )
+    list(APPEND WebCore_LIBRARIES
+        ${LIBGBM_LIBRARIES}
+    )
+endif ()
+
+if (USE_WPE_BACKEND_DRM)
```
Comment (zdobersek) [inline Source/WebCore/PlatformWPE.cmake]: Same for DRM API, this is unused in WebCore so we shouldn't really need the include dir and library flags.

File: Source/cmake/OptionsWPE.cmake [hunk 3]
```diff
@@ -34,6 +34,7 @@ WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_XHR_TIMEOUT PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_WPE_BACKEND_DRM "Whether to enable support for the DRM WPE backend" PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_WPE_BACKEND_DRM_TEGRA "Whether to enable support for the Tegra-specific quirks in the DRM WPE backend" PUBLIC OFF)
 WEBKIT_OPTION_DEFINE(USE_WPE_BACKEND_WAYLAND "Whether to enable support for the Wayland WPE backend" PUBLIC ON)
+WEBKIT_OPTION_DEFINE(USE_WPE_BUFFER_MANAGEMENT_GBM "Whether to enable support for the GBM WPE rendering backend" PUBLIC ON)
```
Comment (zdobersek) [inline Source/cmake/OptionsWPE.cmake]: I think this shouldn't be an option. Instead, we could search for the gbm pkg-config file via find_package(), but not make it a requirement. If found, we would set USE_WPE_BUFFER_MANAGEMENT to 1.

File: Source/WebCore/platform/graphics/GLContext.h [hunk 4]
```diff
@@ -30,9 +30,6 @@
 #if !PLATFORM(GTK) && !PLATFORM(WPE)
 #include "eglplatform.h"
 #else
-#if PLATFORM(GBM)
-#include <gbm.h>
```
Comment (zdobersek) [inline Source/WebCore/platform/graphics/GLContext.h]: This should define __GBM__ to 1 instead, just like gbm.h.

File: Source/WebCore/platform/graphics/wpe/PlatformDisplayWPE.h [hunk 5]
```diff
@@ -28,10 +28,6 @@
 
 #if PLATFORM(WPE)
 
-#if PLATFORM(GBM)
-#include <gbm.h>
```
Comment (zdobersek) [inline Source/WebCore/platform/graphics/wpe/PlatformDisplayWPE.h]: Ditto - this should define __GBM__ to 1 instead, just like gbm.h.

File: Source/cmake/OptionsWPE.cmake [hunk 6]
```diff
@@ -116,10 +120,6 @@ SET_AND_EXPOSE_TO_BUILD(USE_TILED_BACKING_STORE TRUE)
 SET_AND_EXPOSE_TO_BUILD(USE_COORDINATED_GRAPHICS TRUE)
 SET_AND_EXPOSE_TO_BUILD(USE_COORDINATED_GRAPHICS_THREADED TRUE)
 
-if (USE_WPE_BACKEND_DRM OR USE_WPE_BACKEND_WAYLAND)
-    SET_AND_EXPOSE_TO_BUILD(WTF_PLATFORM_GBM TRUE)
```
Comment (zdobersek) [inline Source/cmake/OptionsWPE.cmake]: WTF_PLATFORM_GBM would still be required, but should now just be guarded via USE_WPE_BUFFER_MANAGEMENT_GBM.

**Comment by zdobersek** [comment]
Didn't read your comment before doing another review, sorry.

I agree with the ViewBackendWayland thing, this too should operate only with a buffer management abstraction.

Regarding removing the gbm.h includes, this fails on my system because EGL headers are then included with X11 as the default platform. The most interesting problem that's introduced through this is the replacement of None with 0L because of the X11-specific macro.

File: Source/cmake/OptionsWPE.cmake (hunk 6 above)
Comment (zdobersek) [inline Source/cmake/OptionsWPE.cmake]: In the future, after libWPE (i.e. code under Source/WPE) is spun off into a separate library, there should be some other way to deduce whether the GBM buffer management is supported, either at configure-time or straight at build-time (for instance via a libWPE header that containts configuration-specific macros, or something).

**Comment by blino** [comment]
About the None issue, I got it as well, but assumed it was coming from the recent WPE update.
Thanks for noticing this and the careful review!

**Comment by blino** [comment]
I have kept PLATFORM(GBM) around for the eglplatform.h needs, and removed the outdated DRM/GBM bits from PlatformWPE.cmake

File: Source/cmake/OptionsWPE.cmake (hunk 3 above)
Comment (blino) [inline Source/cmake/OptionsWPE.cmake]: If we do so, this should probably be the same for WPE_BACKEND_DRM and WPE_BACKEND_WAYLAND: build them if detected.
Though, we should still have some way to disable them via cmake options.

Comment (zdobersek) [inline Source/cmake/OptionsWPE.cmake]: OK, let's roll with an option for starters. We can change this when another buffer management API is supported, at the latest.

**Comment by zdobersek** [comment]
This looks and runs good, so if you're fine with it we can do the merge.

**Comment by blino** [comment]
I think this is in a good state for a first step, thanks again for the careful review

**Comment by zdobersek** [comment]
Thanks!

**Pull request merged**